### ISSUE TITLE
Trouble adding environment variables to custom app

### ISSUE THREAD

**Issue by michaelashe** [commercetools/merchant-center-application-kit#1598]
**Describe the issue**
I have a build script and want to put some environment variables here, but i can't do that. When i check at process.env there is only one env var called NODE_ENV and nothing else. How can i add environment variables to build script?

**To Reproduce**
1.create custom app with npx @commercetools-frontend/create-mc-app \
my-new-custom-application-project \
--template starter
2. Go to package.json and add environment variables:
NODE_ENV=production TEST1=var1 TEST2=var2 mc-scripts compile-html
--headers=$(pwd)/headers.staging.json --config=$(pwd)/env.staging.json
--use-local-assets
3. Check process.env. This TEST1 and TEST2 should be able there but there is only NODE_ENV=production

**Expected behavior**
Additional environment variables should be visible in process.env

**Comment by tdeekens** [comment]
Thanks for reaching out. Please specify what version of the Merchant Center Application Kit you are using that helps us narrowing down issues easier.

That said, we do not strip any envrionment variables from the `process.env` but we also use `dotenv` to set them from respective files since recent versions. What do you intend to do with the respective environment variables. Where should they be used. If they are meant to be exposed to the application we would advise to put them into the `env.json` as for instance:

```json
{
   "myOtherEnvVariable": "foo"
}
```

Can be accessed later in the application using `useApplicationContext`:

```js
const myOtherEnvVariable = useApplicationContext(
  (context) => context.environment.myOtherEnvVariable
);
```

Back to the issue at hand. Trying to reproduce on a simple case. Creating a file called `index.js`:

```js`
console.log(process.env);
```

Then running it via `NODE_ENV=production TEST_1=test node index.js` gives the expected output:

```js
{
  // ...other things
  NODE_ENV: 'production',
  TEST_1: 'test',
}
```

This is what I assume you intend to work. Depending on the version of Merchant Center Application Kit being used your start run script should look like `dotenv -- mc-scripts start`. All variables from `.env` will be loaded onto the `process.env`. Instead of specifying it manually I would recommend, if possible, to put them there. You can also have multiple `.env.production` and `.env.staging` files. These would then be merged `dotenv -e .env -e .env.production -- mc-scripts start`. 

Again what you can and should do depends on what you want to achieve (which I am not fully sure of) and what version of AppKit is being used.

**Comment by Sheryorov** [comment]
My deps: 
`  "dependencies": {
    "@commercetools-frontend/actions-global": "16.8.0",
    "@commercetools-frontend/application-components": "16.8.0",
    "@commercetools-frontend/application-shell": "16.8.0",
    "@commercetools-frontend/assets": "16.8.0",
    "@commercetools-frontend/i18n": "16.8.0",
    "@commercetools-frontend/mc-http-server": "16.8.1",
    "@commercetools-frontend/permissions": "16.8.0",
    "@commercetools-frontend/ui-kit": "^10.22.2",
    "@commercetools-uikit/card": "^10.18.4",
    "@commercetools-uikit/collapsible-panel": "^10.18.4",
    "@commercetools-uikit/data-table": "^10.18.4",
    "@commercetools-uikit/flat-button": "10.18.4",
    "@commercetools-uikit/icons": "10.18.4",
    "@commercetools-uikit/spacings": "10.18.4",
    "@commercetools-uikit/text": "10.18.4",
    "apollo-client": "2.6.8",
    "aws4": "^1.9.1",
    "aws4-axios": "^1.12.0",
    "axios": "^0.18.1",
    "crypto": "^1.0.1",
    "deploy-aws-s3-cloudfront": "^3.2.3",
    "dinero.js": "^1.8.1",
    "hex-to-uuid": "^1.0.1",
    "moment": "^2.25.3",
    "prop-types": "15.7.2",
    "react": "16.13.1",
    "react-apollo": "3.1.5",
    "react-dom": "16.13.1",
    "react-intl": "3.4.0",
    "react-redux": "7.2.0",
    "react-router-dom": "5.1.2",
    "redux": "4.0.5"
  },
  "devDependencies": {
    "@commercetools-frontend/jest-preset-mc-app": "16.8.2",
    "@commercetools-frontend/mc-scripts": "^16.12.0",
    "@testing-library/react": "10.0.4",
    "dotenv-cli": "^3.1.0",
    "enzyme": "3.11.0",
    "jest": "26.0.1"
  }`

my env file (.env):
`TEST = 123`

run script:  "dotenv -e .env  -- mc-scripts start"

result: console.log(process.env) => NODE_ENV = Development

**Comment by tdeekens** [comment]
Thanks for the dependencies. 

I the `process.env` is and will be populated within the same node process with all variables specified in the `.env`. I just doule checked.

However, the process might spawn other process, such as the webpack build, this process will have have the environment variables passed automatically. That's intended and how Node.js works and little we can do around it.

Again, where do you want to use the environment variable in the end? Within the React/FE code of the Custom Application?

Then, as mentioned above, I would propose to use the `env.json`. We can use placeholders there to pass things on:

In the `.env`:

```diff
{
+   MY_ENVRIONMENT_VARIABLE: "foo'
}
``` 

Then in the `env.json`:

```diff
{
+   myEnvrionmentVariable: "${env:MY_ENVRIONMENT_VARIABLE}'
}
``` 

The envrionment variable will now be interpolated and be made available through the environment on the application context and can be used there as:

```js
const myEnvrionmentVariable = useApplicationContext(
  (context) => context.environment.myEnvrionmentVariable
);
```

**Comment by Sheryorov** [comment]
Build throwing error: `Error: Missing 'SINGED_URL' specified in config as 'env:SINGED_URL'.
    at substituteEnvVariablePlaceholder (/home/sherzod/workspace/git_projects/orders-custom-app/node_modules/@commercetools-frontend/mc-scripts/node_modules/@commercetools-frontend/mc-html-template/build/utils/substitute-env-variable-placeholders.js:39:11)
    at /home/sherzod/workspace/git_projects/orders-custom-app/node_modules/@commercetools-frontend/mc-scripts/node_modules/@commercetools-frontend/mc-html-template/build/utils/substitute-env-variable-placeholders.js:59:35
    at Array.forEach (<anonymous>)
    at substitutePlaceholders (/home/sherzod/workspace/git_projects/orders-custom-app/node_modules/@commercetools-frontend/mc-scripts/node_modules/@commercetools-frontend/mc-html-template/build/utils/substitute-env-variable-placeholders.js:55:76)
    at /home/sherzod/workspace/git_projects/orders-custom-app/node_modules/@commercetools-frontend/mc-scripts/node_modules/@commercetools-frontend/mc-html-template/build/utils/substitute-env-variable-placeholders.js:68:157
    at Array.map (<anonymous>)
    at substituteEnvVariablePlaceholders (/home/sherzod/workspace/git_projects/orders-custom-app/node_modules/@commercetools-frontend/mc-scripts/node_modules/@commercetools-frontend/mc-html-template/build/utils/substitute-env-variable-placeholders.js:67:76)
    at module.exports (/home/sherzod/workspace/git_projects/orders-custom-app/node_modules/@commercetools-frontend/mc-scripts/node_modules/@commercetools-frontend/mc-html-template/build/load-env.js:28:29)
    at compileHtml (/home/sherzod/workspace/git_projects/orders-custom-app/node_modules/@commercetools-frontend/mc-scripts/node_modules/@commercetools-frontend/mc-html-template/build/compile-html.js:47:15)
    at generateStatic (/home/sherzod/workspace/git_projects/orders-custom-app/node_modules/@commercetools-frontend/mc-scripts/build/compile-html.js:114:26)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! additional-order-info-app@1.0.0 build:staging: `dotenv -e .env.staging -- mc-scripts build && NODE_ENV=production   mc-scripts compile-html --headers=$(pwd)/headers.staging.json --config=$(pwd)/env.staging.json --use-local-assets`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the additional-order-info-app@1.0.0 build:staging script.`

Build script:     "build:staging": "dotenv -e .env.staging -- mc-scripts build && NODE_ENV=production   mc-scripts compile-html --headers=$(pwd)/headers.staging.json --config=$(pwd)/env.staging.json --use-local-assets",

**Comment by tdeekens** [comment]
This is, as far as I read it, cause you have a `SIGNED_URL` in your `env.json` as `${env:SIGNED_URL}` but not in your `.env.staging`.

**Comment by Sheryorov** [comment]
.env.staging
`SINGED_URL =some url
PDF_URL =some rul
SINGED_URL_TOKEN =some token
PDF_URL_TOKEN = some token`

**Comment by Sheryorov** [comment]
.env.json:

>     "signedUrl": "${env:SIGNED_URL}",
>     "signedUrlToken": "${env:SIGNED_URL_TOKEN}",
>     "pdfUrl": "${env:PDF_URL}",
>     "pdfUrlToken": "${env:PDF_URL_TOKEN}"

.env.staging:

> SIGNED_URL = some url
> PDF_URL =  some url
> SIGNED_URL_TOKEN = some token
> PDF_URL_TOKEN = some token

**Comment by Sheryorov** [comment]
it works with `dotenv -e .env.staging -- mc-scripts start` but not working in  build

**Comment by tdeekens** [comment]
Closing due to inactivity and assuming that the solution helps as a result. Feel free to re-open if that was premature.

**Comment by jamesgongqantas** [comment]
I'm having a similar issue, the process.env is not getting the variables defined in .env file. I can use the `useApplicationContext` hook as a workaround. But the hook has to be used in React, for places such as a piece of plain JS file I cannot get the variables from process.env

**Comment by emmenko** [comment]
Are you using `dotenv-cli` to load the environment variables?

```
"start": "dotenv -- mc-scripts start",
```

**Comment by jamesgongqantas** [comment]
> Are you using `dotenv-cli` to load the environment variables?
> 
> ```
> "start": "dotenv -- mc-scripts start",
> ```

yes I did. and I can get the env variables into the env.json file. I just cannot get them from process.env

**Comment by emmenko** [comment]
> I just cannot get them from process.env

I'm not sure what you mean. If you get the variables into the `env.json`, you're good to go.

What are you trying to do exactly and why do you need `process.env`?

**Comment by tdeekens** [comment]
If you intend to access the variables from plain JS I recommend to use the window.app where they should also be globally exposed over process.env. Process.env should only be used in Node.js or in a webpack/script context from my view.

**Comment by jamesgongqantas** [comment]
> > I just cannot get them from process.env
> 
> I'm not sure what you mean. If you get the variables into the `env.json`, you're good to go.
> 
> What are you trying to do exactly and why do you need `process.env`?

if I put variables into `env.json`, I have to get them by using useApplicationContext hook, which means I cannot use it outside a function. I've tried a POC app out from the template and put dotenv in, and I cannot see any other variables other than NODE_ENV in process.env.

**Comment by tdeekens** [comment]
What is directly put into the `env.json` will _not_ automatically be added to the `process.env`. It's what will be pulled into the `window.app` and exposed through the application context, made accessible through the `useApplicationContext` hook.

Again, it depends on what you're trying to achieve. As I understood you want to access the variables through plain JS - not inside a React "context". This means you should be able to use `window.app`. As everything from `env.json` should be there. You don't have to use `process.env` in a browser context.

If you however intend to access the variables in a Node.js context, then you need `process.env`. Then however you have to put the variables into the `.env`-file. This file is what will be sourced by the `dotenv`-cli which will put it onto the `process.env` (the whole nicety of the library itself). You can then reference the values from the `.env` (of which you can have multiple e.g. `.env.staging` and `.env.production`) in the `env.json` through the special syntax of `${env:<NAME>}` (more above and in docs). 

To recap:

1. `env.json`: ends up on `window.app` and in React's application context. Can reference values in `.env`
2. `.env`: ends up on `process.env`

**Comment by emmenko** [comment]
It's still not clear from our side what you're trying to achieve here.

Could you share some code snippet or something that shows what you're trying to do and where exactly (in a react component, in a nodejs environment, etc)?

Thanks

**Comment by jamesgongqantas** [comment]
> What is directly put into the `env.json` will _not_ automatically be added to the `process.env`. It's what will be pulled into the `window.app` and exposed through the application context, made accessible through the `useApplicationContext` hook.
> 
> Again, it depends on what you're trying to achieve. As I understood you want to access the variables through plain JS - not inside a React "context". This means you should be able to use `window.app`. As everything from `env.json` should be there. You don't have to use `process.env` in a browser context.
> 
> If you however intend to access the variables in a Node.js context, then you need `process.env`. Then however you have to put the variables into the `.env`-file. This file is what will be sourced by the `dotenv`-cli which will put it onto the `process.env` (the whole nicety of the library itself). You can then reference the values from the `.env` (of which you can have multiple e.g. `.env.staging` and `.env.production`) in the `env.json` through the special syntax of `${env:<NAME>}` (more above and in docs).
> 
> To recap:
> 
> 1. `env.json`: ends up on `window.app` and in React's application context. Can reference values in `.env`
> 2. `.env`: ends up on `process.env`

sorry to make you guys confused. what I'm trying to do in this case is Node context I guess. In `package.json`, I have `dotenv` and `dotenv-cli` installed.

```
"scripts": {
    "build": "mc-scripts build",
    "start": "dotenv mc-scripts start",
...
```

```
 "dotenv": "^8.2.0",
 "dotenv-cli": "^3.1.0",
```

And this is the `.env` file I have:

```
MC_API_URL=https://mc-api.australia-southeast1.gcp.commercetools.com
FRONTEND_HOST=12localhost:3001
CDN_URL=http://localhost:3001
MC_ENV=development
MC_PROXY_URL=https://mc.australia-southeast1.gcp.commercetools.com
MC_ACCESS_TOKEN= # for schema introspection
```

and my `env.json` file as follows:

```
{
  "applicationName": "custome-app",
  "frontendHost": "${env:FRONTEND_HOST}",
  "mcApiUrl": "${env:MC_API_URL}",
  "location": "au-gcp",
  "env": "${env:MC_ENV}",
  "cdnUrl": "${env:CDN_URL}",
  "servedByProxy": false
}
```
The environment variables in env.json works I believe. However if I do something like the following in `index.js`

`console.log(process.env)`, I can only see NODE_ENV been set.

**Comment by emmenko** [comment]
I think you're confusing a bit how a JavaScript application works and is bundled.

Everything from the application entry point (e.g. `index.js`) is client-side JavaScript, which gets processed and compiled by Webpack. Here, you're not in a Node.js environment, hence you can't do things like `process.env`, using the file system, etc.

To **emulate** the `process.env` part within a client-side (browser) environment, there is a Webpack plugin (`DefinePlugin`) that allows to **inject** variables.
We use that here: https://github.com/commercetools/merchant-center-application-kit/blob/0290743b72b1edee7f6cb8ff90ea894625ef86c0/packages/mc-scripts/src/config/create-webpack-config-for-development.js#L122-L126

This is NOT the `process.env` that you usually see in a Node.js environment. It's literally a hardcoded string/variable that Webpack replaces on compile time.

So doing `console.log(process.env)` doesn't help you.
The `env.json` content is injected into a global variable `window.app`, so if you want to log those values you should do `console.log(window.app)`.

Hope it helps.
